File: ch06/go/exercises.go
package ch06

import "fmt"

func WorstCaseContainsX(str string) int {
	foundX := false
	counter := 0

	for i := range len(str) {
		counter++

		if string(str[i]) == "X" {
			foundX = true
		}
	}

	fmt.Println("Found X: ", foundX)
	fmt.Println("Steps: ", counter)
	return counter
}

func BestAndAverageCaseContainsX(str string) int {
	counter := 0

	for i := range len(str) {
		counter++

		if string(str[i]) == "X" {
			fmt.Println("Found X:  true")
			fmt.Println("Steps: ", counter)
			return counter
		}
	}

	fmt.Println("Found X:  false")
	fmt.Println("Steps: ", counter)
	return counter
}

File: ch06/python/exercises.py
"""
Chapter 06: Exercises
"""


def worst_case_contains_x(string):
    found_x = False
    counter = 0

    for i in range(len(string)):
        counter += 1

        if string[i] == "X":
            found_x = True

    print(f"Found X: {found_x}")
    print(f"Steps: {counter}")
    return counter


print("******** Worst Case Contains X ********")
x_string = "TestX"
steps = worst_case_contains_x(x_string)
assert steps == 5


def best_and_average_case_contains_x(string):
    counter = 0

    for i in range(len(string)):
        counter += 1

        if string[i] == "X":
            print("Found X: True")
            print(f"Steps: {counter}")
            return counter

    print("Found X: False")
    print(f"Steps: {counter}")
    return counter


print("******** Best Case Contains X ********")
x_string = "XTest"
steps = best_and_average_case_contains_x(x_string)
assert steps == 1


print("******** Average Case Contains X ********")
x_string = "TeXst"
steps = best_and_average_case_contains_x(x_string)
assert steps == 3